src/algebra/free_algebra.lean

```diff
@@ -337,4 +339,88 @@ begin
   exact of_id a,
 end
 
+variables {R X}
+
+/--
+Separate an element of the free algebra into its ℕ-graded components.
+
+This is defined as an `alg_hom` to `add_monoid_algebra` so that algebra operators can be moved
+before and after the mapping.
+-/
+noncomputable
+def grades : (free_algebra R X) →ₐ[R] add_monoid_algebra (free_algebra R X) ℕ :=
+lift R $ finsupp.single 1 ∘ ι R
```

Comment:
We have some algebra A (the free algebra), and the question is how to usefully state that A is actually an infinite internal direct sum of submodules A_i. This `grades` map here seems to be a map which does the following: if a is in A, let a_i be its i'th "piece". Then a_i is in A_i and all but finitely many a_i are zero. We can regard the a_i as elements of A. Now the a_i is a finite sequence of elements of A. So we can make them into a polynomial a_i X^i. `grades` sends a to this polynomial. I am surprised that this is a useful concept! But one cool thing about it is that it does seem to be a ring homomorphism, rather than just an R-module homomorphism. The next lemma says that if you evaluate the polynomial at 1 then you get back the element you started with. I've been out of the loop for over a month -- sorry if all this conversation has been had already. Is this the way to do grading of an algebra? It still seems a bit weird to me.

Reply:
I think my choice not to use `polynomial` was deliberate, since I wasn't really trying to invoke the mental model of one (although admittedly I had not spotted that `polynomial A` is defeq to `add_monoid_algebra A ℕ`).

> I am surprised that this is a useful concept!

It seemed to me like the only (and obvious) way to induce a grading from the universal property alone. I suppose though there's no need for it to be the public interface to the grading, it was just simplest to leave it that way, and the fact it was a ring homomorphism seemed cute, even though there's zero reason to need `grades x * grades y` as an alternate spelling of `grades (x * y)`.

> Is this the way to do grading of an algebra?

I assume you mean for a general algebra?

Reply:
> the question is how to usefully state that A is actually an infinite internal direct sum of submodules A_i

I attempted to do this in #4812. The proofs were really quite painful, and I don't think we want to use an API like that until `dfinsupp` is easier to use.